### ISSUE TITLE
Name change?

### ISSUE THREAD

**Issue by jakebolewski** [JuliaGPU/AbstractGPUArray#2]
Could this be `GPUArray`?  Then specific array types could then be `CudaArray`, `CLArray`, `GLArray`, etc which would wrap the underlying buffer types.  `GPUArray` could also be parametrized on the underlying buffer type, ex. `GPUArray{Float64, 2, CLBuffer{Float64}}` so `GPUArray{Float64, 2}` would be the "abstract type".  I feel this would be less verbose and much simpler.

**Comment by SimonDanisch** [comment]
Sure, abstract is awfully long.
I was wondering about how we can encode the location of the buffer in the abstract type. So something like GPUArray{Float32, 2, :OpenCL}. But this is probably not necessary as its encoded one level down in the hierarchy. `GLBuffer <: OGLArray <: GPUArray`
`ex. GPUArray{Float64, 2, CLBuffer{Float64}} so GPUArray{Float64, 2} would be the "abstract type"`
I don't really get your example, to be honest. How does GPUArray has two different sets of parameters for the same thing?
I want to encode quite a lot in the parameters, which seems to make it verbose.
I was thinking, that keeping a copy in ram is a very general thing for GPU arrays, as lookups are so expensive. So that's where I started writing the abstract interface, to give this functionality to both of the main OpenGL arrays (buffers and textures).

**Comment by jakebolewski** [comment]
I think we want to distinguish `Buffer`'s from `Array`'s here.  Buffer's should be low-level and should not be thought of as `Vector`'s.

My thinking was something like this:

```julia
julia> abstract GPUBuffer

julia> immutable CuBuffer{T} <: GPUBuffer
       ptr::Ptr{Void}
       len::Int
       end

julia> immutable CLBuffer{T} <: GPUBuffer
       ptr::Ptr{T}
       len::Int
       end

julia> type GPUArray{T,Dim,Buff<:GPUBuffer} 
       buff::Buff
       size::Dims
       meta
       end
```

That way lots of methods could be written very abstractly against a common "Buffer" interface implemented by the Cuda, OpenCL, OpenGL backends. 

ex: `foo(arr::GPUArray{Float64,2})` would be buffer type agnostic.

If specialization is needed,  the buffer type could be specified `foo{T<:CLFloat}(arr::GPUArray{T, 2, CLBuffer{T}})`.

typealiases could also be added to cut down on the verbositiy:

```julia
typealias CLArray{T,Dim} GPUArray{T,Dim,CLBuffer{T}}
```

**Comment by jakebolewski** [comment]
I should add, you can always add more parameters to the `Buffer` type.  This way they don't leak into the `GPUArray` type.

I'm not really familiar with OpenGL to be honest, what other thing's would you need to parametrize the `GPUBuffer`/`GPUArray` type on other than the element type?

**Comment by SimonDanisch** [comment]
Well, as I said this whole idea started out with my take on cutting down redundancy for working with copies of the array in RAM.
So `setindex!(A::GPUArray{..., RAM})` results in:
`(A[I] = value ; A.ram[I] = value)`
While `GPUArray{..., NORAM}` results in just `A[I] = value`.
After I generalized the problem, this is probably not crucial anymore, and should be encapsulated by another type, like BufferedGPUArray, or what not.
This would actually be rather nice, as it removes this concept from the low-level:
```Julia
immutable BufferedGPUArray{T, D, Buff <: GPUBuffer}
buff::Buff
copy::Array{T, D}
end
```
So +1 to your proposal!

**Comment by jakebolewski** [comment]
@SimonDanisch can you give a couple of `GPUArray` examples with OpenGL buffer/textures?  Really basic stuff, I just want to get a feel for what that would look like.

**Comment by SimonDanisch** [comment]
Wow, totally forgot about this multiple times for 16 days....
So what exactly do you want to see? Some implementations I already have, or how I imagine to make this work with the abstraction in the future?
Here is some example from GLAbstraction:
```Julia
type Texture{T <: SupportedEltypes, ColorDIM, NDIM}
    id::GLuint
    texturetype::GLenum
    pixeltype::GLenum
    internalformat::GLenum
    format::GLenum
    dims::Vector{Int}
    #ram::Array{T, NDIM}
end
```
Now I'd like only to define one simple unsafe getindex and resize for this, and let the abstraction define all other kinds of accesses and also let it do the bound checking etc.

**Comment by SimonDanisch** [comment]
I think I'm starting to get, why I'm not entirely comfortable with your approach.
It might be, because there is a huge difference in functionality between buffers and textures in OpenGL (although, with newer OpenGL API's and more compute function it's diminishing).
From what I know, it seems that compared to OpenGL, the image and buffer types are much more interchangeable in OpenCL.
So as long as you don't need to optimize something, you probably don't care if you use a buffer or an image for the GPUArray.
Most of my functions are exclusively defined for either GLBuffer or GLTexture. So I don't profit much from the abstraction GPUArray{Buffer}. In the contrary, it's much more cumbersome to use than `Texture <: GPUArray` & `Buffer <: GPUArray`.
I actually just want to have a simple way, to define all the array functions for my opengl types, without changing much for my inner functionality.
And probably define some map operations and such. 
But I have to think about this... Probably I'm missing something, which can make this work nicely!

**Comment by jakebolewski** [comment]
@SimonDanisch why is it more cumbersome to use?  If you are special casing `getindex` / `setindex` for instance for `GLBuffer` and `GLTexture`, how is that different than writing two implementations one for `GPUArray{GLBuffer}`, `GPUArray{GLTexture}`?  Functionality that builds on top of `getindex` / `setindex`  could be parameterized on an abstract `GPUArray` type (not parameterized on the specific buffer type).

I want to experiment with this idea some more.  I've been waiting to update OpenCL until 0.4 comes out but that might be a while so it would be good to get this working if only to sanely interface with the CL / CUDA math libraries.

**Comment by SimonDanisch** [comment]
>I want to experiment with this idea some more. I've been waiting to update OpenCL until 0.4 comes out but that might be a while so it would be good to get this working if only to sanely interface with the CL / CUDA math libraries.

Same here! I just got compute shaders running, which feel a little like OpenCL for OpenGL, which gives us the possibility to treat them much more like arrays.

Getindex/setindex on GPUarray{GLTexture} wouldn't be a great problem. I just wonder about the user facing side. If the user thinks in buffers and textures GPUArray{GLBuffer} won't be a nice abstraction for him, as he now as do deal with a more complex type without obvious benefits. 
So having GPUArray abstract and simply inheriting from it would be easier for that user.
My plan so far is this kind of interface:
```Julia
#Package GLAbstraction
#implement the interface (hope it's the same situation for other gpu arrays)
# at least in OpenGL, for an n dimensional array, there is only 1 native get/setindex
gpu_getindex!(GLBuffer{T, 1}, i::UnitRange) = ...
gpu_getindex!(GLTexture{T, 1}, i::UnitRange) = ...
gpu_getindex!(GLTexture{T, 2}, i::UnitRange, j::UnitRange) = ...
gpu_getindex!(GLTexture{T, 3}, i::UnitRange, j::UnitRange, k::UnitRange) = ...

#Package GPUArray

**Comment by jakebolewski** [comment]
You are right that this approach may be too complicated for no real benefit.  If we have `GPUArray` as the abstract type and then have `CLArray`, `CuArray` as subtypes of `GPUArray` then we can always move to what I am proposing in the future without much code churn.

**Comment by SimonDanisch** [comment]
Yes this sounds sensible! What would be your use cases, by the way?

2015-03-18 21:31 GMT+01:00 Jake Bolewski <notifications@github.com>:

**Comment by jakebolewski** [comment]
In the near term, mostly  calling BLAS / FFT / LAPACK functionality with GPU arrays and getting out rows / columns of the resulting arrays on the GPU.
